### PR TITLE
Adding csv subscription management.

### LINKED ISSUE #30: Enhance publishing to accept a csv blob for restricted audience
```
restrictedAudience": {
  "uploadedCsvUri": "yourCsvUriGoesHere"
},
```

### PR DESCRIPTION
**Pull request by kwoodson**
Adding csv support for managing subscriptions.

fixes #30

### CODE REVIEW

**Comment by kwoodson** [comment]
Still requires testing.  I was able to download/upload csv with changes.

Tests:
- test add subscription (sort of tested)
- test remove subscription
- test new option for csv uri (undocumented option)
- test add image

File: src/go/azure-publish/pkg/publisher/api/v20171031/types.go [hunk 1]
```diff
@@ -76,7 +76,8 @@ type Plan struct {
 }
 
 type RestrictedAudience struct {
-	ManualEntries []ManualEntry `json:"manualEntries"`
+	ManualEntries  []ManualEntry `json:"manualEntries"`
+	UploadedCsvUri string        `json:"uploadedCsvUri"`
```
Comment (kwoodson) [inline src/go/azure-publish/pkg/publisher/api/v20171031/types.go]: New option provided by MSFT.  Need to test to ensure it works.

File: src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+func (s *Subscription) String() string {
+	return fmt.Sprintf("subscriptionId,%v,%v", s.Id, s.Description)
+}
+
+func fetchSubscriptionCSV(offerName, sku, containerName string) ([]Subscription, error) {
+	var subscriptions []Subscription
+	csvfile := fmt.Sprintf("%v_%v.csv", offerName, sku)
+
+	blobClient := azcli.storage.GetBlobService()
```
Comment (mjudeikis) [inline src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go]: where `azcli` comes from?

Comment (mjudeikis) [reply src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go]: Can we have follow-up issue to remove azcli and rework this all package to be `pkg` `cmd` based? I already see this is growing a little bit more than simple golang application, and some package based structure would be useful here.

Comment (kwoodson) [reply src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go]: We do have `cmd` and `pkgz do exist.  The `azcli` comes from here:

https://github.com/openshift/azure-misc/blob/master/azure-publish/cmd/azure-publish/main.go#L24-L25

I'm definitely open to suggestions on improving this code.

File: src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
+package main
```
Comment (jim-minter) [inline src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go]: @kwoodson am supposing this file should not be present in this PR (remnant of a rebase after the src/go move?)

Comment (kwoodson) [reply src/go/azure-publish/cmd/azure-publish/addcustomersubscription.go]: Will combine this with addsubscription.go.

File: azure-publish/pkg/publisher/api/v20171031/types.go [hunk 4]
```diff
@@ -57,26 +57,33 @@ type Offer struct {
...
+	Regions                    []string                `json:"regions"`
 }
 
 type RestrictedAudience struct {
-	ManualEntries []ManualEntry `json:"manualEntries"`
+	ManualEntries  []ManualEntry `json:"manualEntries"`
+	UploadedCsvUri string        `json:"uploadedCsvUri"`
+}
+
+type VirtualMachinePricingV2 struct {
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/pkg/publisher/api/v20171031/types.go]: Yes and this was a change recently made by MSFT to the portal and when not present was complaining about it.

File: azure-publish/cmd/azure-publish/removeimage.go [hunk 5]
```diff
@@ -5,7 +5,7 @@ import (
 	"fmt"
 	"strings"
 
-	"github.com/openshift/azure-misc/src/go/azure-publish/pkg/publisher/api/v20171031"
+	"github.com/openshift/azure-misc/azure-publish/pkg/publisher/api/v20171031"
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/removeimage.go]: https://github.com/openshift/azure-misc/issues/85

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 6]
```diff
@@ -4,9 +4,10 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/openshift/azure-misc/src/go/azure-publish/pkg/publisher/api/v20171031"
+	"github.com/openshift/azure-misc/azure-publish/pkg/publisher/api/v20171031"
 )
 
+// This adds a subscription to the Preview Subscription list
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Fixed.

File: azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+	return csvurl, nil
+}
+
+func addCustomerSubscriptionToOffer(offerName, sku, container string, offer *v20171031.OfferWrapper) (bool, error) {
+	planID := sku
+
+	subs, err := fetchSubscriptionCSV(offerName, sku, container)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addcustomersubscription.go]: Thanks.  Not sure how this panic was here.  Updated.

File: azure-publish/cmd/azure-publish/addimage.go [hunk 8]
```diff
@@ -62,7 +62,12 @@ func addimage(resourceGroup string, imageName string, offerName string, containe
 		return err
 	}
 
-	if !updated {
+	updated2, err := addCustomerSubscriptionToOffer(offerName, sku, subscriptionContainer, offer)
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addimage.go]: We can remove this.  This allows you to add a subscription id and an image in the same call.  Simpler if we remove.

File: azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+func removeCustomerSubscription(id string, subs []Subscription) []Subscription {
+	removeIdx := -1
+	for idx, sub := range subs {
+		if sub.Id == id {
+			removeIdx = idx
+			break
+		}
+	}
+
+	subs = append(subs[:removeIdx], subs[removeIdx+1:]...)
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addcustomersubscription.go]: Will remove only when index is not -1.

File: azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+	readCloser, err := blob.Get(nil)
+	if err != nil {
+		return subscriptions, err
+	}
+
+	reader := csv.NewReader(readCloser)
+
+	// initial read to skip headers
+	_, err = reader.Read()
+	if err == io.EOF {
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addcustomersubscription.go]: removed.

File: azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+		fields, err := reader.Read()
+
+		if err == io.EOF {
+			break // end of data
+		} else if err != nil {
+			return nil, err
+		}
+
+		subscriptions = append(subscriptions, Subscription{
+			Id:          fields[1],
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addcustomersubscription.go]: Added.

File: azure-publish/cmd/azure-publish/addcustomersubscription.go [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func addCustomerSubscriptionToOffer(offerName, sku, container string, offer *v20171031.OfferWrapper) (bool, error) {
+	planID := sku
+
+	subs, err := fetchSubscriptionCSV(offerName, sku, container)
+	if err != nil {
+		panic(err)
+	}
+	subs, modified := appendCustomerSubscription(*subscription, *description, subs)
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addcustomersubscription.go]: Agreed.  Will pass them.

File: azure-publish/pkg/publisher/api/v20171031/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply azure-publish/pkg/publisher/api/v20171031/types.go]: I mean - if VirtualMachinePricing and VirtualMachinePricingV2 are identical definitions, why define them twice?

File: azure-publish/cmd/azure-publish/addimage.go [hunk 13]
```diff
@@ -46,7 +46,7 @@ func addImageToOffer(offer *v20171031.OfferWrapper, image *compute.Image, contai
 	return true, nil
 }
 
-func addimage(resourceGroup string, imageName string, offerName string, container string, sku string) error {
+func addimage(resourceGroup, imageName, offerName, container, subscriptionContainer, sku string) error {
```
Comment (jim-minter) [inline azure-publish/cmd/azure-publish/addimage.go]: no need to pass subscriptionContainer - you're not doing anything with it

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 14]
```diff
@@ -38,3 +50,187 @@ func addsubscription(subscription string, offerName string) error {
...
+
+func fetchSubscriptionCSV(offerName, sku, containerName string) ([]Subscription, error) {
+	var subscriptions []Subscription
+	csvfile := fmt.Sprintf("%v_%v.csv", offerName, sku)
+
+	blobClient := azcli.storage.GetBlobService()
+	blob := blobClient.GetContainerReference(containerName).GetBlobReference(csvfile)
+	readCloser, err := blob.Get(nil)
+	if err != nil {
+		return subscriptions, err
```
Comment (jim-minter) [inline azure-publish/cmd/azure-publish/addsubscription.go]: return nil, err

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 15]
```diff
@@ -38,3 +50,187 @@ func addsubscription(subscription string, offerName string) error {
...
+	}
+
```
Comment (jim-minter) [inline azure-publish/cmd/azure-publish/addsubscription.go]: defer readCloser.Close()

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 16]
```diff
@@ -38,3 +50,187 @@ func addsubscription(subscription string, offerName string) error {
...
+	subs = append(subs, Subscription{
+		Type:        "subscriptionId",
+		Id:          id,
+		Description: description,
+	})
+
+	return subs, true
+}
+
+func removeCustomerSubscription(id string, subs []Subscription) []Subscription {
```
Comment (jim-minter) [inline azure-publish/cmd/azure-publish/addsubscription.go]: it is worth noting that this algorithm will not remove the subscription correctly if it appears multiple times (which I saw the other day as the uploaded csv was corrupt).  The following is shorter and more robust:
```
func removeCustomerSubscription(id string, subs []Subscription) []Subscription {
	newSubs := make([]Subscription, 0, len(subs))
	for _, sub := range subs {
		if sub.Id != id {
			newSubs = append(newSubs, sub)
		}
	}
 	return newSubs
}
```

File: azure-publish/pkg/publisher/api/v20171031/client.go [hunk 17]
```diff
@@ -71,12 +72,15 @@ func (c *Client) do(method string, url string, inData interface{}, header http.H
...
 	resp, err := http.DefaultClient.Do(req)
+
 	if err != nil {
 		return err
 	}
 	defer resp.Body.Close()
 
 	if resp.StatusCode/100 != 2 {
+		res, _ := httputil.DumpResponse(resp, true)
+		fmt.Println(string(res))
```
Comment (jim-minter) [inline azure-publish/pkg/publisher/api/v20171031/client.go]: is this intended to be here?

File: azure-publish/cmd/azure-publish/addsubscription.go (hunk 14 above)
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Done

File: azure-publish/cmd/azure-publish/addsubscription.go (hunk 15 above)
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Added.

File: azure-publish/cmd/azure-publish/addsubscription.go (hunk 16 above)
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Thanks, much cleaner!

File: azure-publish/pkg/publisher/api/v20171031/client.go (hunk 17 above)
Comment (kwoodson) [reply azure-publish/pkg/publisher/api/v20171031/client.go]: Yes, when I received a 400 error from the client there was no information to assist.  By adding this debug statement I was able to see that the csv was invalid.

Is this acceptable?

File: azure-publish/pkg/publisher/api/v20171031/client.go [hunk 18]
```diff
@@ -71,12 +72,15 @@ func (c *Client) do(method string, url string, inData interface{}, header http.H
 	req.Header = header
 
 	resp, err := http.DefaultClient.Do(req)
+
 	if err != nil {
 		return err
 	}
 	defer resp.Body.Close()
 
 	if resp.StatusCode/100 != 2 {
```
Comment (pweil-) [inline azure-publish/pkg/publisher/api/v20171031/client.go]: are you doing this in this fashion for a particular reason (as opposed to using a constant like https://golang.org/src/net/http/status.go)?  Just stuck out as strange to me, it certainly works and isn't part of this PR so I'm not asking it to change now.

Comment (kwoodson) [reply azure-publish/pkg/publisher/api/v20171031/client.go]: This was some magic that Jim put in when he refactored my initial work.

File: azure-publish/pkg/publisher/api/v20171031/types.go [hunk 19]
```diff
@@ -76,7 +77,8 @@ type Plan struct {
 }
 
 type RestrictedAudience struct {
-	ManualEntries []ManualEntry `json:"manualEntries"`
+	ManualEntries  []ManualEntry `json:"manualEntries"`
+	UploadedCsvUri string        `json:"uploadedCsvUri"`
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/pkg/publisher/api/v20171031/types.go]: Fixed.

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 20]
```diff
@@ -1,12 +1,24 @@
...
+	"encoding/csv"
 	"fmt"
+	"io"
+	"time"
 
+	azstorage "github.com/Azure/azure-sdk-for-go/storage"
 	"github.com/openshift/azure-misc/azure-publish/pkg/publisher/api/v20171031"
 )
 
+type Subscription struct {
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Fixed

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 21]
```diff
@@ -1,12 +1,24 @@
...
+	Type        string
+	Id          string
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Fixed.

**Comment by kwoodson** [comment]
Tested this morning by adding new subscription and publish.

File: azure-publish/pkg/publisher/api/v20171031/client.go (hunk 18 above)
Comment (jim-minter) [reply azure-publish/pkg/publisher/api/v20171031/client.go]: iirc the api varies between 200, 201 and 202 in what it returns and we treat them equivalently

**Comment by jim-minter** [comment]
just 2 nits then lgtm

File: azure-publish/cmd/azure-publish/addsubscription.go [hunk 22]
```diff
@@ -38,3 +51,185 @@ func addsubscription(subscription string, offerName string) error {
...
+
+func addCustomerSubscriptionToOffer(offerName, sku, container, subscription, description string, offer *v20171031.OfferWrapper) (bool, error) {
+	planID := sku
+
+	subs, err := fetchSubscriptionCSV(offerName, sku, container)
+	if err != nil {
+		return false, err
+	}
+	if subs == nil {
+		return false, fmt.Errorf("unable to fetch subscription csv")
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply azure-publish/cmd/azure-publish/addsubscription.go]: Yes, if the blob is not in the storage account it returns:
```
panic: storage: service returned error: StatusCode=404, ErrorCode=BlobNotFound, ErrorMessage=The specified blob does not exist.
```

**Comment by jim-minter** [comment]
tests passed, manually squash merging

**Pull request merged**